File: src/main/java/src/gestorAplicacion/personas/Estudiante.java
package gestorAplicacion.personas;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import gestorAplicacion.Calendario.*;
import gestorAplicacion.Calendario.Horario.dias;
import gestorAplicacion.Calendario.Materia.tipo;
import java.io.Serializable;

public class Estudiante extends Persona implements Serializable{
	
	//atributos
	private double promedio;
	private boolean calificacionAsignada;
	private boolean fueBecado;
	public ArrayList<Materia> materias_inscritas;
	private double porcentajeDeAvance;
	private boolean fallaHorario;
	protected ArrayList<Materia> materias_cursadas;
	private int creditosInscritos = 0;
	private ArrayList<Materia> intentoMaterias;
	private ArrayList<Profesor> profesoreInscritos;
		
	//constructor
	public Estudiante(String nombre, int ID, String Email, boolean fueBecado, ArrayList<Materia> materias_cursadas){
		super (nombre, ID, Email);
		this.fallaHorario = false;
		this.calificacionAsignada = false;
		this.fueBecado = fueBecado;
		this.promedio =  0.0;
		materias_inscritas = new ArrayList<Materia>();
		this.materias_cursadas=materias_cursadas;
		intentoMaterias = new ArrayList<Materia>();
		profesoreInscritos= new ArrayList<Profesor>();

	}
	public Estudiante(String nombre, int ID, String Email, boolean fueBecado){
		super (nombre, ID, Email);
		this.fallaHorario = false;
		this.fueBecado = fueBecado;
		this.promedio =  0.0;
		materias_inscritas = new ArrayList<Materia>();
		materias_cursadas = new ArrayList<Materia>();
		intentoMaterias = new ArrayList<Materia>();
		profesoreInscritos= new ArrayList<Profesor>();
	}
	//Metodos get y set
	
	public void setProfesoresInscritos(ArrayList<Profesor> profesor) {
		profesoreInscritos = profesor;
	}
	
	public ArrayList<Profesor> getProfesoresInscritos(){
		return profesoreInscritos;
	}
	
	public ArrayList<Materia> getMaterias_Inscritas(){
		return materias_inscritas;
	}
	
	public void setMaterias_Inscritas(ArrayList<Materia> materiasInscritas) {
		materias_inscritas = materiasInscritas;
	}
	
	
	public void setfallaHorario(boolean fallaHorario) {
		this.fallaHorario= fallaHorario;
	}
	public boolean getfallaHorario() {
		return fallaHorario;
	}
	
	
	public boolean getFueBecado() {
        return fueBecado;
    }
	public void setFueBecado(boolean fueBecado) {
        this.fueBecado = fueBecado;
    }
	
	
	public double getPorcentajeDeAvance() {
	 	return porcentajeDeAvance;
 	}
	public void setPorcentajeDeAvance(double PorcentajeDeAvance) {
	 	porcentajeDeAvance = PorcentajeDeAvance;
 	}	 
	 public ArrayList<Materia> getMateriasCursadas() {
	        return materias_cursadas;
	    }
	 public void setMateriasCursadas(ArrayList<Materia> materias_cursadas) {
			this.materias_cursadas=materias_cursadas;
	    }
	 
	 public boolean getCalificacionAsignada() {
			return calificacionAsignada;
		}
		public void setCalificacionAsignada(boolean calificacionAsignada) {
			this.calificacionAsignada = calificacionAsignada;
		}
	 
	 public void setCreditosInscritos(int creditosInscritos) {
			this.creditosInscritos= creditosInscritos;
		}
		public int getCreditosInscritos() {
			return creditosInscritos;
		}
	 
		
	//metodos de la clase
		public boolean inscribirMateria(String nombreMateria, ArrayList<Materia> materiasDisponibles) {
		    boolean tieneFundamentacion = false;
		    boolean falloPrerrequsito = false;
		    int intentoCreditos = 0;

		    for (Materia materia : materiasDisponibles) {
		        if (materia.getNombre().equals(nombreMateria)) {
		            if (materias_inscritas.contains(materia)) {
		                return false;
		            }
		            Materia prerrequisito = materia.getPrerrequisito();
		            boolean tienePrerrequisito = false;
		            if (prerrequisito != null) {
		                for (Materia materiaCursada : materias_cursadas) {
		                    if (materiaCursada.getNombre().equals(prerrequisito.getNombre())) {
		                        tienePrerrequisito = true;
		                        break;
		                    }
		                }
		            }
		            if (prerrequisito == null || tienePrerrequisito) {
		                intentoMaterias.add(materia);
		                break;
		            } else {
		                falloPrerrequsito = true;
		            } 
		        }
		    }

		    for (Materia m : intentoMaterias) {
		        intentoCreditos += m.getCreditos();
		        if (m.getTipo() == tipo.fundamentacion) {
		            tieneFundamentacion = true;
		        }
		    }

		    if (intentoCreditos >= 6 && tieneFundamentacion) {
		        for(Materia ma: intentoMaterias) {
		            profesoreInscritos.add(ma.getProfesor());
		            ma.inscribirEstudiante(this);
		        }
		        setMaterias_Inscritas(intentoMaterias);
		        setCreditosInscritos(intentoCreditos);
		    }

		    return !falloPrerrequsito;
		}
	        
	
	public void retirarMateria(Materia Materia) {
		materias_inscritas.remove(Materia);
	}
	
	
	public void aplicarBeca (Estudiante estudiante) {
		Beca.estudiantes.add(this);
	}
	
	
	public double calcularPromedio() {
	    double finalScore = 0.0;
	    int numMaterias = materias_inscritas.size();
	    for (Materia materia : materias_inscritas) {
	        double materiaScore = 0.0;
	        int numTareas = materia.getTareasDeMateria().size();
	        for (Tarea tarea : materia.getTareasDeMateria()) {
	            materiaScore += tarea.getGrade(this);
	        }
	        if (numTareas > 0) {
	            materiaScore /= numTareas;
	        }
	        finalScore += materiaScore;
	    }
	    if (numMaterias > 0) {
	        finalScore /= numMaterias;
	    }
	    return Math.round(finalScore * 100.0) / 100.0;
	}
	
	//funcionalidades 1. Detectar problemas con franja horaria
	public boolean compararHorario(ArrayList<Materia> materias_inscritas){
		ArrayList<Materia>MateriasError = new ArrayList<Materia>();
		fallaHorario = false;
		for(int i = 0; i < materias_inscritas.size(); i++ ) {
			Horario horario = materias_inscritas.get(i).getHorario();
			String hora1 = horario.getHora_inicio();
			String hora2 = horario.getHora_Fin();
			ArrayList<dias> dia1 = horario.getDia();
			
			for (int j = i + 1; j < materias_inscritas.size(); j++) {
				Horario horario2 = materias_inscritas.get(j).getHorario();
				String hora3 = horario2.getHora_inicio();
				String hora4 = horario2.getHora_Fin();
				ArrayList<dias> dia2 = horario2.getDia();
			
				for (int k = 0; k < dia1.size(); k++) {
					dias dia11 = dia1.get(k);
					for (int h = 0; h < dia2.size(); h++) {
						dias dia22= dia2.get(h);
					if (dia11 == dia22 && !MateriasError.contains(materias_inscritas.get(i)))  {
						if (hora1 == hora3 || hora2 == hora4) {
							 fallaHorario=true; 
							 MateriasError.add(materias_inscritas.get(i));
							 MateriasError.add(materias_inscritas.get(j));
							}
						}
					}
				}
			}
		}
		return fallaHorario;
	}
	// Funcionalidad 2: Sugerir Horario y materia
	public void sugerirHorario(boolean fallaHorario) {
	    ArrayList<Materia> materiasSugeridas = new ArrayList<Materia>();
	    if (fallaHorario) {
	        for (Materia materia : materias_inscritas) {
	            boolean conflicto = false;
	            for (Materia materiaSugerida : materiasSugeridas) {
	                
	                ArrayList<dias> diasMateria = materia.getHorario().getDia();
	                ArrayList<dias> diasMateriaSugerida = materiaSugerida.getHorario().getDia();
	                boolean falloDias = false;
	                for (dias diaMateria : diasMateria) {
	                    if (diasMateriaSugerida.contains(diaMateria)) {
	                    	falloDias = true;
	                        break;
	                    }
	                }
	                
	                if (falloDias &&
	                    (materia.getHorario().getHora_inicio().equals(materiaSugerida.getHorario().getHora_inicio()) ||
	                     materia.getHorario().getHora_Fin().equals(materiaSugerida.getHorario().getHora_Fin()))) {
	                    conflicto = true;
	                    break;
	                }
	            }
	            if (!conflicto) {
	                materiasSugeridas.add(materia);
	            }
	        }
	      this.materias_inscritas = materiasSugeridas;
	    }
	}
	
	//Funcionalidad 3: Crear un horario en base a las materias vistas
	public void sugerirMaterias(ArrayList<Materia> materiasDisponibles) {  
		ArrayList<Materia> materiasRecomendadas = new ArrayList<Materia>();
		for (Materia materia : materias_cursadas) {
			
			if(materia.getNombre().equalsIgnoreCase(materiasDisponibles.get(0).getNombre()) && !(materias_cursadas.contains(materiasDisponibles.get(0)))){
				materiasRecomendadas.add(materiasDisponibles.get(0));
				profesoreInscritos.add(materiasDisponibles.get(0).getProfesor());
				materiasDisponibles.get(0).inscribirEstudiante(this);
			}
			
			else if(materia.getNombre().equalsIgnoreCase(materiasDisponibles.get(0).getNombre()) && !(materias_cursadas.contains(materiasDisponibles.get(1)))) {
				materiasRecomendadas.add(materiasDisponibles.get(1));
				profesoreInscritos.add(materiasDisponibles.get(1).getProfesor());
				materiasDisponibles.get(1).inscribirEstudiante(this);
			}
			else if(materia.getNombre().equalsIgnoreCase(materiasDisponibles.get(1).getNombre()) && !(materias_cursadas.contains(materiasDisponibles.get(2)))) {
				materiasRecomendadas.add(materiasDisponibles.get(2));
				profesoreInscritos.add(materiasDisponibles.get(2).getProfesor());
				materiasDisponibles.get(2).inscribirEstudiante(this);
			}
		
			if (materia.getNombre().equalsIgnoreCase(materiasDisponibles.get(3).getNombre()) && !(materias_cursadas.contains(materiasDisponibles.get(3)))) {
				materiasRecomendadas.add(materiasDisponibles.get(3));
				profesoreInscritos.add(materiasDisponibles.get(3).getProfesor());
				materiasDisponibles.get(3).inscribirEstudiante(this);
				if(compararHorario(materiasRecomendadas)) {
					materiasRecomendadas.remove(materiasDisponibles.get(3));
					profesoreInscritos.remove(materiasDisponibles.get(3).getProfesor());
					materiasDisponibles.remove(3).inscribirEstudiante(this);
				}
			}
			
			else if (materia.getNombre().equalsIgnoreCase(materiasDisponibles.get(3).getNombre()) && !(materias_cursadas.contains(materiasDisponibles.get(4)))) {
				materiasRecomendadas.add(materiasDisponibles.get(4));
				profesoreInscritos.add(materiasDisponibles.get(4).getProfesor());
				materiasDisponibles.get(4).inscribirEstudiante(this);
				if(compararHorario(materiasRecomendadas)) {
					materiasRecomendadas.remove(materiasDisponibles.get(4));
					profesoreInscritos.remove(materiasDisponibles.get(4).getProfesor());
					materiasDisponibles.remove(4).inscribirEstudiante(this);
				}
			}
			else if(materia.getNombre().equalsIgnoreCase(materiasDisponibles.get(4).getNombre()) && !(materias_cursadas.contains(materiasDisponibles.get(5)))) {
				materiasRecomendadas.add(materiasDisponibles.get(5));
				profesoreInscritos.add(materiasDisponibles.get(5).getProfesor());
				materiasDisponibles.get(5).inscribirEstudiante(this);
				if(compararHorario(materiasRecomendadas)) {
					materiasRecomendadas.remove(materiasDisponibles.get(5));
					profesoreInscritos.remove(materiasDisponibles.get(5).getProfesor());
					materiasDisponibles.remove(5).inscribirEstudiante(this);
				}
			}
		}
		
		for (int i = 6; i<9;i++) {
			materiasRecomendadas.add(materiasDisponibles.get(i));
			profesoreInscritos.add(materiasDisponibles.get(i).getProfesor());
			materiasDisponibles.get(i).inscribirEstudiante(this);
			if(compararHorario(materiasRecomendadas)) {
				materiasRecomendadas.remove(materiasDisponibles.get(i));
				profesoreInscritos.remove(materiasDisponibles.get(i).getProfesor());
				materiasDisponibles.remove(i).inscribirEstudiante(this);
			}
		}
		this.materias_inscritas = materiasRecomendadas;
	}
	
	
	public String toString () {
		return "Estudiante: " + getNombre();
		}
	
	public void calcularPorcetajeAvance() {
		double porcentajeAvance = (1.0/9)*materias_cursadas.size();
		setPorcentajeDeAvance(Math.round(porcentajeAvance * 100.0) );
	}
	
	
	
	
	
}

		
	
	

File: src python/gestorAplicacion.personas/Estudiante.py
from typing import List
from gestorAplicacion.Calendario import Horario, Materia
from gestorAplicacion.Personas.Persona import Persona
import pickle

class Estudiante(Persona):
    def __init__(self, nombre: str, ID: int, Email: str, fueBecado: bool, materias_cursadas: List[Materia]):
        super().__init__(nombre, ID, Email)
        self.promedio = 0.0
        self.calificacionAsignada = False
        self.fueBecado = fueBecado
        self.materias_inscritas = []
        self.porcentajeDeAvance = 0.0
        self.fallaHorario = False
        self.materias_cursadas = materias_cursadas
        self.creditosInscritos = 0
        self.intentoMaterias = []
        self.profesoreInscritos = []

    def __init__(self, nombre, ID, Email, fueBecado):
        super().__init__(nombre, ID, Email)
        self.fallaHorario = False
        self.fueBecado = fueBecado
        self.promedio = 0.0
        self.materias_inscritas = []
        self.materias_cursadas = []
        self.intentoMaterias = []
        self.profesoreInscritos = []

    def setProfesoresInscritos(profesor):
        global profesoreInscritos
        profesoreInscritos = profesor

    def getProfesoresInscritos():
        return profesoreInscritos

    def getMaterias_Inscritas():
        return materias_inscritas

    def setMaterias_Inscritas(materiasInscritas):
        global materias_inscritas
        materias_inscritas = materiasInscritas

    def setfallaHorario(fallaHorario):
        global fallaHorario
        fallaHorario = fallaHorario

    def getfallaHorario():
        return fallaHorario

    def getFueBecado():
        return fueBecado

    def setFueBecado(fueBecado):
        global fueBecado
        fueBecado = fueBecado

    def getPorcentajeDeAvance():
        return porcentajeDeAvance

    def setPorcentajeDeAvance(PorcentajeDeAvance):
        global porcentajeDeAvance
        porcentajeDeAvance = PorcentajeDeAvance

    def getMateriasCursadas():
        return materias_cursadas

    def setMateriasCursadas(materias_cursadas):
        global materias_cursadas
        materias_cursadas = materias_cursadas

    def getCalificacionAsignada():
        return calificacionAsignada

    def setCalificacionAsignada(calificacionAsignada):
        global calificacionAsignada
        calificacionAsignada = calificacionAsignada

    def setCreditosInscritos(creditosInscritos):
        global creditosInscritos
        creditosInscritos = creditosInscritos

    def getCreditosInscritos():
        return creditosInscritos

    def inscribirMateria(nombreMateria, materiasDisponibles):
        global materias_inscritas, profesoreInscritos
        tieneFundamentacion = False
        falloPrerrequsito = False
        intentoCreditos = 0
        intentoMaterias = []
        for materia in materiasDisponibles:
            if materia.getNombre() == nombreMateria:
                if materia in materias_inscritas:
                    return False
                prerrequisito = materia.getPrerrequisito()
                tienePrerrequisito = False
                if prerrequisito != None:
                    for materiaCursada in materias_cursadas:
                        if materiaCursada.getNombre() == prerrequisito.getNombre():
                            tienePrerrequisito = True
                            break
                if prerrequisito == None or tienePrerrequisito:
                    intentoMaterias.append(materia)
                    break
                else:
                    falloPrerrequsito = True
        for m in intentoMaterias:
            intentoCreditos += m.getCreditos()
            if m.getTipo() == tipo.fundamentacion:
                tieneFundamentacion = True
        if intentoCreditos >= 6 and tieneFundamentacion:
            for ma in intentoMaterias:
                profesoreInscritos.append(ma.getProfesor())
                ma.inscribirEstudiante(this)
            setMaterias_Inscritas(intentoMaterias)
            setCreditosInscritos(intentoCreditos)
        return not falloPrerrequsito

    def retirarMateria(Materia):
        global materias_inscritas
        materias_inscritas.remove(Materia)

    def aplicarBeca(estudiante):
        Beca.estudiantes.append(this)

    def calcularPromedio():
        finalScore = 0.0
        numMaterias = len(materias_inscritas)
        for materia in materias_inscritas:
            materiaScore = 0.0
            numTareas = len(materia.getTareasDeMateria())
            for tarea in materia.getTareasDeMateria():
                materiaScore += tarea.getGrade(this)
            if numTareas > 0:
                materiaScore /= numTareas
            finalScore += materiaScore
        if numMaterias > 0:
            finalScore /= numMaterias
        return round(finalScore * 100.0) / 100.0

    def compararHorario(materias_inscritas):
        MateriasError = []
        fallaHorario = False
        for i in range(len(materias_inscritas)):
            horario = materias_inscritas[i].getHorario()
            hora1 = horario.getHora_inicio()
            hora2 = horario.getHora_Fin()
            dia1 = horario.getDia()
            for j in range(i + 1, len(materias_inscritas)):
                horario2 = materias_inscritas[j].getHorario()
                hora3 = horario2.getHora_inicio()
                hora4 = horario2.getHora_Fin()
                dia2 = horario2.getDia()
                for k in range(len(dia1)):
                    dia11 = dia1[k]
                    for h in range(len(dia2)):
                        dia22 = dia2[h]
                        if dia11 == dia22 and materias_inscritas[i] not in MateriasError:
                            if hora1 == hora3 or hora2 == hora4:
                                fallaHorario = True
                                MateriasError.append(materias_inscritas[i])
                                MateriasError.append(materias_inscritas[j])
        return fallaHorario

    def sugerirHorario(fallaHorario):
        materiasSugeridas = []
        if fallaHorario:
            for materia in materias_inscritas:
                conflicto = False
                for materiaSugerida in materiasSugeridas:
                    diasMateria = materia.getHorario().getDia()
                    diasMateriaSugerida = materiaSugerida.getHorario().getDia()
                    falloDias = False
                    for diaMateria in diasMateria:
                        if diaMateria in diasMateriaSugerida:
                            falloDias = True
                            break
                    if falloDias and (materia.getHorario().getHora_inicio() == materiaSugerida.getHorario().getHora_inicio() or materia.getHorario().getHora_Fin() == materiaSugerida.getHorario().getHora_Fin()):
                        conflicto = True
                        break
                if not conflicto:
                    materiasSugeridas.append(materia)
            materias_inscritas = materiasSugeridas

    def sugerirMaterias(materiasDisponibles):
        materiasRecomendadas = []
        for materia in materias_cursadas:
            if materia.getNombre().lower() == materiasDisponibles[0].getNombre().lower() and materiasDisponibles[0] not in materias_cursadas:
                materiasRecomendadas.append(materiasDisponibles[0])
                profesoreInscritos.append(materiasDisponibles[0].getProfesor())
                materiasDisponibles[0].inscribirEstudiante(self)
            elif materia.getNombre().lower() == materiasDisponibles[0].getNombre().lower() and materiasDisponibles[1] not in materias_cursadas:
                materiasRecomendadas.append(materiasDisponibles[1])
                profesoreInscritos.append(materiasDisponibles[1].getProfesor())
                materiasDisponibles[1].inscribirEstudiante(self)
            elif materia.getNombre().lower() == materiasDisponibles[1].getNombre().lower() and materiasDisponibles[2] not in materias_cursadas:
                materiasRecomendadas.append(materiasDisponibles[2])
                profesoreInscritos.append(materiasDisponibles[2].getProfesor())
                materiasDisponibles[2].inscribirEstudiante(self)
            if materia.getNombre().lower() == materiasDisponibles[3].getNombre().lower() and materiasDisponibles[3] not in materias_cursadas:
                materiasRecomendadas.append(materiasDisponibles[3])
                profesoreInscritos.append(materiasDisponibles[3].getProfesor())
                materiasDisponibles[3].inscribirEstudiante(self)
                if compararHorario(materiasRecomendadas):
                    materiasRecomendadas.remove(materiasDisponibles[3])
                    profesoreInscritos.remove(materiasDisponibles[3].getProfesor())
                    materiasDisponibles[3].inscribirEstudiante(self)
            elif materia.getNombre().lower() == materiasDisponibles[3].getNombre().lower() and materiasDisponibles[4] not in materias_cursadas:
                materiasRecomendadas.append(materiasDisponibles[4])
                profesoreInscritos.append(materiasDisponibles[4].getProfesor())
                materiasDisponibles[4].inscribirEstudiante(self)
                if compararHorario(materiasRecomendadas):
                    materiasRecomendadas.remove(materiasDisponibles[4])
                    profesoreInscritos.remove(materiasDisponibles[4].getProfesor())
                    materiasDisponibles[4].inscribirEstudiante(self)
            elif materia.getNombre().lower() == materiasDisponibles[4].getNombre().lower() and materiasDisponibles[5] not in materias_cursadas:
                materiasRecomendadas.append(materiasDisponibles[5])
                profesoreInscritos.append(materiasDisponibles[5].getProfesor())
                materiasDisponibles[5].inscribirEstudiante(self)
                if compararHorario(materiasRecomendadas):
                    materiasRecomendadas.remove(materiasDisponibles[5])
                    profesoreInscritos.remove(materiasDisponibles[5].getProfesor())
                    materiasDisponibles[5].inscribirEstudiante(self)
        for i in range(6, 9):
            materiasRecomendadas.append(materiasDisponibles[i])
            profesoreInscritos.append(materiasDisponibles[i].getProfesor())
            materiasDisponibles[i].inscribirEstudiante(self)
            if compararHorario(materiasRecomendadas):
                materiasRecomendadas.remove(materiasDisponibles[i])
                profesoreInscritos.remove(materiasDisponibles[i].getProfesor())
                materiasDisponibles[i].inscribirEstudiante(self)
        self.materias_inscritas = materiasRecomendadas

    def __str__():
        return "Estudiante: " + getNombre()

    def calcularPorcentajeAvance():
        porcentajeAvance = (1.0/9)*len(materias_cursadas)
        setPorcentajeDeAvance(round(porcentajeAvance * 100.0))
